topology/probes/ovs_of.go

```diff
@@ -0,0 +1,531 @@
+	/* #nosec */
+	command := exec.CommandContext(ctx, com, args...)
+	out, err := command.StdoutPipe()
+	command.Stderr = command.Stdout
+	if err != nil {
+		return out, err
+	}
+	err = command.Start()
+	return out, err
+}
+
+// LaunchOnSwitch launches a command on a given switch
+func LaunchOnSwitch(cmd []string) (string, error) {
+	bytes, err := executor.ExecCommand(cmd[0], cmd[1:]...)
+	if err == nil {
+		return string(bytes), nil
+	}
+	return "", err
+}
+
+// LaunchContinuousOnSwitch launches  a stream producing command on a given switch
+func LaunchContinuousOnSwitch(ctx context.Context, cmd []string) (<-chan string, error) {
+	var cout = make(chan string, 10)
+	out, err := executor.ExecCommandPipe(ctx, cmd[0], cmd[1:]...)
+	if err != nil {
+		close(cout)
+		return cout, err
+	}
+	reader := bufio.NewReader(out)
+	var line string
+
+	go func() {
+		for {
+			line, err = reader.ReadString('\n')
+			if err == io.EOF {
+				close(cout)
+				break
+			} else if err != nil {
+				logging.GetLogger().Errorf("IO Error on command %s: %s", cmd[0], err.Error())
+			} else {
+				cout <- line
+			}
+		}
+	}()
+
+	return cout, nil
+}
+
+func countElements(filter string) int {
+	if len(filter) == 0 {
+		return 1
+	}
+	elts := strings.Split(filter, ",")
+	l := len(elts) + 1
+	for _, elt := range elts {
+		if strings.HasPrefix(elt, "priority=") {
+			l = l - 1
+			break
+		}
+	}
+	return l
+}
+
+// completeRule completes the rule by looking at it again but with dump-flows. This gives back more elements such as priority.
+func completeRule(o *OvsOfProbe, event *Event) error {
+	oldrule := event.Rule
+	bridge := event.Bridge
+	// We want exactly n+1 items where n was the number of items in old filters. The reason is that now
+	// the priority is provided. Another approach would be to use the shortest filter as it is the more generic
+	expected := countElements(oldrule.Filter)
+	filter := makeFilter(oldrule)
+	command, err1 := o.makeCommand([]string{"ovs-ofctl", "dump-flows"}, bridge, filter)
+	if err1 != nil {
+		return err1
+	}
+	lines, err := LaunchOnSwitch(command)
+	if err != nil {
+		return fmt.Errorf("Cannot launch ovs-ofctl dump-flows on %s@%s with filter %s: %s", bridge, o.Host, filter, err.Error())
+	}
+	done := false
+	for _, line := range strings.Split(lines, "\n") {
+		rule, err2 := parseRule(line)
+		if err2 == nil && countElements(rule.Filter) == expected && oldrule.Cookie == rule.Cookie {
+			if done {
+				logging.GetLogger().Errorf("Multiple completion for rule on %s@%s with filter %s", bridge, o.Host, filter)
+			}
+			oldrule.Filter = rule.Filter
+			oldrule.Actions = rule.Actions
+			done = true
+		}
+	}
+	extractPriority(oldrule)
+	if !done {
+		return fmt.Errorf("Cannot complete rule on %s@%s with filter %s", bridge, o.Host, filter)
+	}
+	return nil
+}
+
+// addRule adds a rule to the graph and links it to the bridge.
+func addRule(rule *Rule, probe *BridgeOfProbe) {
```

Comment:
Done in https://github.com/skydive-project/skydive/pull/296/commits/f293bca068724639314b1791b85ae94ec1c5ed63 but I am not convinced. The point is that these are low level functions and there is no interface to use at the caller site. For example monitor directly uses the layout of the probe, so it will not use an interface instead of the actual type for the receiver type when it calls addRule or delRule. As a result, there is no gain of abstraction converting functions to methods (even if there is no loss in performance because the type is known). The use of methods without interface is misleading.